### PR TITLE
Be more concrete

### PR DESCRIPTION
**Pull request by glwagner**
Builds off @vchuravy's work in #261.

Preliminary design of user API for concrete models. Probably most of the time the user will have to specify boundary conditions prior to building the model. We'll have to come up with good syntax for that. What we have is ok, but can be improved.

The checkpointing tests currently fail, I think because they attempt to change the types of the some of the model fields.

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
Please don't delete the `glwagner:concrete` branch. We can fix the BC API issue from #272 and reopen the PR.

File: test/test_boundary_conditions.jl [hunk 1]
```diff
@@ -45,8 +56,13 @@ end
 function test_flux_budget(arch, TF, field_name)
     N, κ, Lz = 16, 1, 0.7
 
+    bottom_flux = TF(0.3)
+    flux_bc = BoundaryCondition(Flux, bottom_flux)
+    modelbcs = BoundaryConditions(field_name, :z, :bottom, flux_bc)
```
Comment (glwagner) [inline test/test_boundary_conditions.jl]: @ali-ramadhan see here for how to create `BoundaryConditions` with one non-default boundary conditions.

File: test/test_regression.jl [hunk 2]
```diff
@@ -177,25 +176,25 @@ function run_rayleigh_benard_regression_test(arch)
...
     model = Model(
          arch = arch,
             N = (Nx, Ny, Nz),
             L = (Lx, Ly, Lz),
             ν = ν,
             κ = κ,
           eos = LinearEquationOfState(βT=1., βS=0.),
-    constants = PlanetaryConstants(g=1., f=0.)
+    constants = PlanetaryConstants(g=1., f=0.),
+          bcs = BoundaryConditions(T=FieldBoundaryConditions(z=ZBoundaryConditions(
```
Comment (glwagner) [inline test/test_regression.jl]: @ali-ramadhan see here for how to set boundary conditions in `z` on a particular field.

**Comment by glwagner** [comment]
As noted in my description, I did write a minimal API (a few keyword constructors, and also one convenience function to instantiate `ModelBoundaryConditions` with one non-default boundary condition. I use the API in the tests. It does need documentation.

A better API could perhaps introduce a `BoundaryConditionsWizard` type that allows users to use the syntax

```julia
wizard = BoundaryConditionsWizard()
wizard.T.z.top = BoundaryCondition(Flux, 1.0)

# some time later...

model = Model(; boundary_conditions=wizard)
```

`BoundaryConditionsWizard` would just be a 100% abstract implementation of `ModelBoundaryConditions` (basically, what we had before). Then we could write a function that converts a `Wizard` to real boundary conditions with concrete types in the `Model` constructor. 

That's one idea. @vchuravy, any other ideas?
